vpr/src/analytical_place/ap_mass_report.cpp

```diff
@@ -0,0 +1,542 @@
+/**
+ * @file
+ * @author  Alex Singer
+ * @date    May 2025
+ * @brief   Implementation of the AP mass report generator.
+ */
+
+#include "ap_mass_report.h"
+#include <cmath>
+#include <fstream>
+#include <iomanip>
+#include <ostream>
+#include <string>
+#include <vector>
+#include "ap_netlist.h"
+#include "ap_netlist_fwd.h"
+#include "globals.h"
+#include "logic_types.h"
+#include "physical_types.h"
+#include "physical_types_util.h"
+#include "primitive_vector.h"
+#include "vpr_context.h"
+#include "vpr_error.h"
+#include "vtr_time.h"
+#include "vtr_vector.h"
+
+namespace {
+
+/**
+ * @brief A node in a printing tree.
+ */
+struct PrintingTreeNode {
+    /// @brief The name of this node. This will be printed when the tree is printed.
+    std::string name;
+    /// @brief The children of this node in the tree.
+    std::vector<PrintingTreeNode> children;
+};
+
+/**
+ * @brief A printing tree. This tree contains basic information which will be
+ *        used to print data in a good-looking form.
+ */
+struct PrintingTree {
+    /// @brief The root node of the tree.
+    PrintingTreeNode root;
+};
+
+/**
+ * @brief Print the given tree node.
+ *
+ *  @param node
+ *      The node to print.
+ *  @param os
+ *      The output file stream to print the node to.
+ *  @param prefix
+ *      The prefix that all children of this node will print before their names.
+ */
+void print_tree_node(const PrintingTreeNode& node,
```

Comment:
I'm not sure whether it's a common convention or not, but in other parts of VTR, the "recur" suffix is added to the names of recursive functions.